cxx-qt-lib/src/qurl.rs

```diff
@@ -0,0 +1,175 @@
+    fn unique_ptr_qurl_get(this: *const MaybeUninit<*mut c_void>) -> *const QUrl;
+    #[link_name = "cxxqt1$unique_ptr$qurl$release"]
+    fn unique_ptr_qurl_release(this: *mut MaybeUninit<*mut c_void>) -> *mut QUrl;
+    #[link_name = "cxxqt1$unique_ptr$qurl$drop"]
+    fn unique_ptr_qurl_drop(this: *mut MaybeUninit<*mut c_void>);
+}
+
+// Safety: TODO
+unsafe impl UniquePtrTarget for QUrl {
+    #[doc(hidden)]
+    fn __typename(f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str("QUrl")
+    }
+
+    #[doc(hidden)]
+    fn __null() -> MaybeUninit<*mut c_void> {
+        let mut repr = MaybeUninit::uninit();
+        unsafe {
+            unique_ptr_qurl_null(&mut repr);
+        }
+        repr
+    }
+
+    #[doc(hidden)]
+    unsafe fn __raw(raw: *mut Self) -> MaybeUninit<*mut c_void> {
+        let mut repr = MaybeUninit::uninit();
+        unique_ptr_qurl_raw(&mut repr, raw);
+        repr
+    }
+
+    #[doc(hidden)]
+    unsafe fn __get(repr: MaybeUninit<*mut c_void>) -> *const Self {
+        unique_ptr_qurl_get(&repr)
+    }
+
+    #[doc(hidden)]
+    unsafe fn __release(mut repr: MaybeUninit<*mut c_void>) -> *mut Self {
+        unique_ptr_qurl_release(&mut repr)
+    }
+
+    #[doc(hidden)]
+    unsafe fn __drop(mut repr: MaybeUninit<*mut c_void>) {
+        unique_ptr_qurl_drop(&mut repr)
+    }
+}
+
+pub struct Url {
+    // Note that once map_qt_value is removed later, this can become private again
+    #[doc(hidden)]
+    pub inner: cxx::UniquePtr<QUrl>,
+}
+
+impl Url {
+    pub fn from_qurl(qurl: &QUrl) -> Self {
+        Self {
+            // Safety: TODO
+            inner: unsafe {
+                let mut ptr = MaybeUninit::<cxx::UniquePtr<QUrl>>::zeroed();
+                qurl_init_from_qurl(&mut ptr, qurl);
+                ptr.assume_init()
+            },
+        }
+    }
+
+    pub fn from_string(s: String) -> Self {
+        Self {
+            // Safety: TODO
+            inner: unsafe {
+                let mut ptr = MaybeUninit::<cxx::UniquePtr<QUrl>>::zeroed();
+                qurl_init_from_string(&mut ptr, s.as_ptr(), s.len());
+                ptr.assume_init()
+            },
+        }
+    }
+
+    // TODO: other QUrl methods
+    //
+    // fragment: Option<String>,
+    // host: Option<String>,
+    // password: Option<String>,
+    // path: Option<String>,
+    // port: i32,
+    // query: Option<String>,
+    // scheme: Option<String>,
+    // userName: Option<String>,
+
+    pub fn string(&self) -> String {
+        let mut s = String::new();
+        unsafe { qurl_to_rust_string(&self.inner, &mut s) };
+        s
+    }
+}
+
+impl crate::ToUniquePtr for Url {
+    type CppType = QUrl;
+
+    /// Retrieve the UniquePtr to the Qt QUrl of this Rust Url
+    /// so that this object can be passed back to C++.
+    fn to_unique_ptr(self) -> cxx::UniquePtr<QUrl> {
+        self.inner
+    }
+}
+
+impl From<&QUrl> for Url {
+    fn from(qurl: &QUrl) -> Self {
+        qurl.to_rust()
```

Comment:
When #9 is implemented, maybe we should reconsider whether to use a From<>, or a custom ToRust Trait.
Currently, the to_rust method is somewhat unnecessary, as we could just use .into() and implement the conversion directly in the trait impl.

Reply:
Right, maybe note that in #9 or somewhere :-) Although I think i was thinking of just removing implementing a trait and instead just making the macro expansion put `to_rust()` where it needs to.